### PR TITLE
Adjust React's setState to work in more edge cases

### PR DESCRIPTION
**Pull request by ericanderson**
Fixes DefinitelyTyped/DefinitelyTyped#22230
Fixes DefinitelyTyped/DefinitelyTyped#18365

### CODE REVIEW

**Comment by ericanderson** [comment]
@vsiao Mind pushing through?

**Comment by ericanderson** [comment]
@RyanCavanaugh Can you help get this pushed through? I'm trying to fix a regression I caused yesterday

**Comment by johnnyreilly** [comment]
Merged - thanks!

File: types/react/index.d.ts [hunk 1]
```diff
@@ -282,8 +282,9 @@ declare namespace React {
 
         // We MUST keep setState() as a unified signature because it allows proper checking of the method return type.
         // See: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18365#issuecomment-351013257
+        // Also, the ` | S` allows intellisense to not be dumbisense
```
Comment (natehitchcock) [inline types/react/index.d.ts]: Solid comment, haha

**Pull request merged**